Document the Release Process (#188)
veraison/go-cose: release-management.md

# go-cose Release Management

## Overview

This document describes [go-cose][go-cose] project release management, which includes release criteria, versioning, supported releases, and supported upgrades.

The go-cose project maintainers strive to provide a stable go-lang implementation for interacting with [COSE][ietf-cose] constructs.
Stable implies appropriate and measured changes to the library assuring consumers have the necessary functionality to interact with COSE objects.
If you or your project require added functionality, or bug fixes, please open an issue or create a pull request.
The project welcomes all contributions from adding functionality, implementing testing to security reviews.
Please see [here](https://github.com/veraison#contributing) for how to contribute.

> 

Reviewer: Please also state, somewhere in this document, where and how we want the external users to understand what is the current stable `go-cose` release ?

Reviewer: is it through the REAMDE.md or somewhere else?

Author: I was trying to avoid having to update the readme every time we do a release, and leave it to the [github releases](https://github.com/veraison/go-cose/releases) page. However, the [current readme](https://github.com/veraison/go-cose?tab=readme-ov-file#project-status) does state the current release, so we could continue to update that, or remove it and simply say, check the releases page for the most current.

Do you think we need to update both?

Reviewer: I think, it is sensible to add a permanent reference in README.md where to find the latest `go-cose` release!

A new major or minor release will not have an automated build/release posted until the branch reaches alpha quality.

- all versions use a preface of `v`
- `X` is the [Major](#major-releases) version
- `Y` is the [Minor](#minor-releases) version
- `Z` is the [Patch](#patch-releases) version
- _Optional_ `-alpha.n` | `-rc.n` [pre-release](#pre-release) version
  - Each incremental alpha or rc build will bump the suffix (`n`) number.
  - It's not expected to have more than 9 alphas or rcs.
The suffix will be a single digit.
  - If > 9 builds do occur, the format will simply use two digit indicators (`v1.0.0-alpha.10`)

_**Note**: Pre-releases will NOT use the github pre-release flag._

## Branch Management

To meet the projects stability goals, go-cose does not currently operate with multiple feature branches.
All active development happens in main.
For each release, a branch is created for servicing, following the versioning name.
`v1.0.0-alpha-1` would have an associated [v1.0.0-alpha.1](https://github.com/veraison/go-cose/tree/v1.0.0-alpha.1) branch.
All version and branch names are lower case.

> 

Reviewer: Do we want to maintain all the Release Branches..?

Author: I believe the answer should be yes. Even if something is unsupported, or even has a bug/issue, a consumer may be dependent upon it until they can make changes, or take time to take a newer version.

### Major Releases

The go-cose project will issue major releases when substantial functionality has changed, including:

> - New capabilities consumers may want to opt-into, rather than automatically get as a result of getting the latest minor release or patch.

Reviewer: Line 48 is not clear to me.

New capabilities consumers may want to opt-into.

I do not think, new functionality can be obtained by getting the latest minor release... I am confused what the text is intending to state?

Author: Thanks, Yogesh.
Please take a look at the updated changes.

New and unexpected capabilities _can_ be an attack vector.
Making explicit opt-in decisions to upgrade, or possibly downgrade should be the choice of a project consuming go-cose.
- Breaking Changes, where projects should make an implicit decision to upgrade, and adjust for breaking changes.

Each major release will go through one or more `-alpha.n` and `-rc.n` pre-release phases.

### Minor Releases

> The go-cose project will issue minor releases when incremental improvements, or bug fixes are added to existing functionality in a [Major Release](#major-releases).

Reviewer: ```suggestion
The go-cose project will issue minor releases when incremental improvements, or bug fixes are added to existing functionality.
```

Principals of a patch release:

- Should be a "safe bet" to upgrade to.
- No breaking changes.
- No feature or surface area changes.
- A "bug fix" that may be a breaking change may require a major release.
- Applicable fixes, including security fixes, may be cherry-picked from main into the latest supported minor release-X.Y branches.
- Patch releases are cut from a release-X.Y.Z branch.

Each patch release will go through one or more `-alpha.n` and `-rc.n` pre-release phases.

### Pre-Release

As builds of main become stable, and a pending release is planned, a pre-release build will be made.
Pre-releases go through one or more `-alpha.n` releases, followed by one or more incremental `-rc.n` releases.

- **alpha.n:** `X.Y.Z-alpha.n`
  - alpha release, cut from the branch where development occurs.
To minimize branch management, no additional branches are maintained for each incremental release.
  - Considered an unstable release which should only be used for early development purposes.
  - Released incrementally until no additional issues and prs are made against the release.

>   - Once no triaged issues or prs are scoped to the release, an rc is cut.

Reviewer 2: ```suggestion
  - Once no triaged issues or pull requests (prs) are scoped to the release, a release candidate (rc) is cut.
```

  - To minimize confusion, and the risk of an alpha being widely deployed, alpha branches and released binaries may be removed at the discretion, and a [two-thirds supermajority][super-majority] vote, of the maintainers.
Maintainers will create an Issue, and vote upon it for transparency to the decision to remove a release and/or branch.

>   - Not supported

Reviewer 2: ?

Author: Can I get a wee bit more about which part is the question? 
This section is focused on deleting alphas and rcs, as a forcing function (when needed) to an actual release. 
We know that orgs wind up depending on pre-release builds, and we could break someone by removing a pre-release build, so I'm just putting some process around this.

Reviewer: I think, the comment is for line 93, What is the context of `-Not supported` ?

Author: I can add a link to the ["support"](https://github.com/veraison/go-cose/pull/188/files#diff-0f9f71fb66352a76730712a4204600abd7a6f46b400b4345886ab61cf30c45b4R105) section

Author: Added a link to the not-supported line. Pease unresolve if it's something else.

- **rc.n:** `X.Y.Z-rc.n`
  - Released as needed before a final version is released
  - Bugfixes on new features only as reported through usage
  - An rc is not expected to revert to an alpha release.
  - Once no triaged issues or prs are scoped to the release, an final version is cut.
  - A release candidate will typically have at least two weeks of bake time, providing the community time to provide feedback.

>   - Release candidate are cut from the branch where the work is done.

Reviewer: ```suggestion
  - Release candidate is cut from the branch where the work is done.
```

  - To minimize confusion, and the risk of an rc being widely deployed, rc branches and released binaries may be removed at the discretion, and a [two-thirds supermajority][super-majority] vote, of the maintainers.
Maintainers will create an Issue, and vote upon it for transparency to the decision to remove a release and/or branch.

>   - Not supported

Author: same as [above](https://github.com/veraison/go-cose/pull/188#discussion_r1674277901)

Support will be provided best effort by the maintainers via GitHub issues and pull requests from the community.

The go-cose maintainers expect users to stay up-to-date with the versions of go-cose release they use in production, but understand that it may take time to upgrade.
We expect users to be running approximately the latest patch release of a given minor release and encourage users to upgrade as soon as possible.

While pre-releases may be deleted at the discretion of the maintainers, all Major, Minor and Patch releases should be maintained.
Only in extreme circumstances, as agreed upon by a [two-thirds supermajority][super-majority] of the maintainers, shall a release be removed.

Applicable fixes, including security fixes, may be cherry-picked into the release branch, depending on severity and feasibility.
Patch releases are cut from that branch as needed.

## Security Reviews

The go-cose library is an sdk around underlying crypto libraries, tailored to COSE scenarios.
The go-cose library does not implement cryptographic functionality, reducing the potential risk.

> To assure go-cose had the proper baseline, two [security reviews](./reports) were conducted prior to the [v1.0.0](https://github.com/veraison/go-cose/releases/tag/v1.0.0) release

Reviewer 2: ```suggestion
```

Reviewer 2: No need for history in a policy doc

Author: The purpose was to reference the existing security reviews, to provide visibility and confidence in how we manage our policy. We have, and will do security reviews, but will not do them for all releases.
How do we create visibility to reviews, so it doesn't look like we're dismissing the need?
